components/sup/src/manager/mod.rs

```diff
@@ -934,138 +1096,213 @@ impl Manager {
+        user_config_watcher: Arc<RwLock<UserConfigWatcher>>,
+        updater: Arc<Mutex<ServiceUpdater>>,
+        busy_services: Arc<Mutex<HashSet<PackageIdent>>>,
+        reconcile_services: Arc<AtomicBool>,
+    ) -> impl Future<Item = (), Error = ()> {
+        // JW TODO: Update service rumor to remove service from
+        // cluster
+        // TODO (CM): But only if we're not going down for a restart.
+        let ident = service.spec_ident.clone();
+        let stop_it = service.stop().then(move |_| {
+            user_config_watcher
                 .write()
-                .expect("Services lock is poisoned!");
-            mem::swap(services.deref_mut(), &mut svcs);
-        }
+                .expect("Watcher lock poisoned")
+                .remove(&service);
+            updater
+                .lock()
+                .expect("Updater lock poisoned")
+                .remove(&service);
+            Ok(())
+        });
+        Self::wrap_async_service_operation(ident, busy_services, reconcile_services, stop_it)
+    }
+
+    /// Wrap a future that starts, stops, or restarts a service with
+    /// logic that marks that service as "busy" for the duration of
+    /// the process.
+    ///
+    /// This allows us to postpone taking additional action on
+    /// services until they're done with what they're doing.  (For
+    /// example, consider stopping a service that takes 10 seconds to
+    /// shut down (including post-stop hook execution), but then
+    /// executing `hab svc start SERVICE` 2 seconds into that 10
+    /// seconds.)
+    ///
+    /// As more service operations (e.g., hooks) become asynchronous,
+    /// we'll need to wrap those operations in this logic to ensure
+    /// consistent operation.
+    fn wrap_async_service_operation<F>(
+        ident: PackageIdent,
+        busy_services: Arc<Mutex<HashSet<PackageIdent>>>,
+        reconcile_services: Arc<AtomicBool>,
+        fut: F,
+    ) -> impl Future<Item = (), Error = ()>
+    where
+        F: IntoFuture<Item = (), Error = ()>,
+    {
+        // TODO (CM): can't wait for the Pinning API :(
+        let busy_services_2 = busy_services.clone();
+        let ident_2 = ident.clone();
+
+        future::lazy(move || {
+            trace!(
+                "Flagging '{:?}' as busy, pending an asynchronous operation",
+                ident
+            );
+            busy_services
+                .lock()
+                .expect("busy_services lock is poisoned")
+                .insert(ident);
+            Ok(())
+        })
+        .and_then(|_| fut)
+        .and_then(move |_| {
+            trace!(
+                "Removing 'busy' flag for '{:?}'; asynchronous operation over",
+                ident_2
+            );
+            busy_services_2
+                .lock()
+                .expect("busy_services lock is poisoned")
+                .remove(&ident_2);
+            reconcile_services.store(true, Ordering::Relaxed);
```

Comment:
I don't think it's going to be a performance hit or anything, but I think that `Relaxed` is really all that we need here. Every time we're finishing one of these futures, we're unconditionally setting the flag to `true`, and the manager looks at it each time through and only takes action if it's `true` (while simultaneously setting it back to `false` in the process). 

If a bunch of futures finish, it doesn't matter what order they set the flag to `true`. Similarly, when responding to the flag, a strict ordering isn't necessary.

Reply:
I think in terms of overall correctness, there's no issue, but if we don't have any requirements on ordering, I think there's a chance we could do the reconciliation later than we'd prefer since re-ordering of instructions could cause the check to occur earlier than the set even though the order of the operations in reality was reversed. I could be wrong; I'm hardly an expert on this.

My thinking here is just that `SeqCst` lets the compiler deviate from our logical expectations as humans as little as possible, and so I see it as the default unless performance requirements demand otherwise. I don't actually expect that it'll matter much either way here.

Reply:
Even if there was re-ordering, the worst that would happen is that we'd just catch it the next time through the main loop.

I chose `Relaxed` because that seems to most accurately reflect the scenario we have.

The better approach would probably be to wrap all this up in a type, so that this ordering business is completely an internal detail, and represented in a single place. What are your thoughts on that?

Reply:
> Even if there was re-ordering, the worst that would happen is that we'd just catch it the next time through the main loop.

Yeah, that's what I meant by "a chance we could do the reconciliation later than we'd prefer". Since users may notice that delay, it seems the more appropriate thing to try to optimize.

> The better approach would probably be to wrap all this up in a type, so that this ordering business is completely an internal detail, and represented in a single place. What are your thoughts on that?

It would be nice not to have to say the ordering everywhere. The only downside is that it wouldn't be appropriate for some of the ordering types where the gets and sets are *supposed* to use different values. But I would be in favor of `hcore::AtomicXXX` types that all used `SeqCst` always internally.

Reply:
I was specifically thinking of making something like `ReconciliationFlag(Arc<AtomicBool>)` that would have method for setting and getting; that way any implementation logic is consolidated right there, and not scattered around the manager. Then, we can change the ordering to whatever we want, and be sure that we're doing things properly for that particular type. I wasn't thinking of making a general `AtomicSeqCstBool`.